### PR TITLE
BUG: Any dtype should call `square` on `arr ** 2`

### LINKED ISSUE #29388: BUG: `power` called instead of `square` in `__array_ufunc__` for structured arrays
### Describe the issue:

`__array_ufunc__` catches `array ** 2` as `power` and not `square` for structured arrays. The following works well -

```py
import numpy


class DummySubclass(numpy.ndarray):
    def __array_ufunc__(
        self,
        ufunc,
        method,
        *inputs,
        **kwargs,
    ):
        print(ufunc)
        if ufunc is numpy.power:
            print("numpy.power called")
        elif ufunc is numpy.square:
            print("numpy.square called")


d = DummySubclass(shape=(1,1), dtype=float)
d**2
# output:
# <ufunc 'square'>
# numpy.square called
```

but changing `dtype` to that of a structured array (see below) sets `ufunc` as `power`. The behavior for structured and non-structures arrays was identical before NumPy `2.3.0`.

### Reproduce the code example:

```python
import numpy


class DummySubclass(numpy.ndarray):
    def __array_ufunc__(
        self,
        ufunc,
        method,
        *inputs,
        **kwargs,
    ):
        print(ufunc)
        if ufunc is numpy.power:
            print("numpy.power called")
        elif ufunc is numpy.square:
            print("numpy.square called")


d = DummySubclass(shape=(1,1), dtype=[('x', '<f8'), ('y', '<f8'), ('z', '<f8'), ('t', '<f8')])
d**2
```

### Error message:

```shell
<ufunc 'power'>
numpy.power called
```

### Python and NumPy Versions:

2.3.1 (can confirm that the bug was introduced in 2.3.0)
3.13.5 (main, Jun 11 2025, 15:36:57) [Clang 17.0.0 (clang-1700.0.13.3)]

### Runtime Environment:

```
[{'numpy_version': '2.3.1',
  'python': '3.13.5 (main, Jun 11 2025, 15:36:57) [Clang 17.0.0 '
            '(clang-1700.0.13.3)]',
  'uname': uname_result(system='Darwin', node='Saranshs-MacBook-Pro-3.local', release='24.5.0', version='Darwin Kernel Version 24.5.0: Tue Apr 22 19:54:29 PDT 2025; root:xnu-11417.121.6~2/RELEASE_ARM64_T6030', machine='arm64')},
 {'simd_extensions': {'baseline': ['NEON', 'NEON_FP16', 'NEON_VFPV4', 'ASIMD'],
                      'found': ['ASIMDHP'],
                      'not_found': ['ASIMDFHM']}}]
```

### Context for the issue:

This is a silent bug which is messing up a lot of calculations in our work (and I am assuming that it is doing the exact same in other codebases without people actually noticing it).

I will be more than happy to fix it in NumPy if someone could point me to the right direction. Thanks!

### PR DESCRIPTION
**Pull request by MaanasArora**
Fixes #29388.

The simplification in [gh-27901](https://github.com/numpy/numpy/pull/27901) seems to have removed special casing for squares, not considering the differing behavior of `np.power(x, 2)` and `np.square(x)` on structured dtypes.

There may be more edge cases to consider--I am looking! Also adding a test. Thanks.

### CODE REVIEW

File: numpy/_core/src/multiarray/number.c [hunk 1]
```diff
@@ -363,7 +366,12 @@ fast_scalar_power(PyObject *o1, PyObject *o2, int inplace, PyObject **result)
     }
 
     PyArrayObject *a1 = (PyArrayObject *)o1;
-    if (!(PyArray_ISFLOAT(a1) || PyArray_ISCOMPLEX(a1))) {
+    if (PyArray_ISOBJECT(a1)) {
```
Comment (seberg) [inline numpy/_core/src/multiarray/number.c:367]: Thanks, yeah this seems right/safe to me.  I am not sure I see why `**2` confuses a custom `__array_ufunc__`, but I don't think this is an issue (and probably restores old behavior).  (If we had a bigger reason to keep it as is, I would be tempted to ask downstream to fix it there.)

Curious, does `**0.5` always end up with `sqrt()`?

---

(Maybe the nicer solution for at least `**2` would actually be to make the object `square` implementation use `**2` internally then this special case wouldn't even be needed.  But no priority so if, let's do that as a follow-up.)

**Review by seberg — APPROVED** [review]
Anyway, looks right, will put it in a bit. Thanks a lot!

File: numpy/_core/src/multiarray/number.c [hunk 2]
```diff
@@ -363,7 +366,12 @@ fast_scalar_power(PyObject *o1, PyObject *o2, int inplace, PyObject **result)
...
+        return 1;
+    }
+    if (!is_square && !PyArray_ISFLOAT(a1) && !PyArray_ISCOMPLEX(a1)) {
```
Comment (seberg) [inline numpy/_core/src/multiarray/number.c]: Actually, sorry should have looked closer.  This is either square or sqrt.  Maybe we can just remove this branch also for `sqrt`?
Or would that change behavior drastically from the original version?  (I think I would lean towards that.

(If not, it may be nice to just use `fastop != nops.square` rather than introducing a new variable, but nitpicking.)

Comment (MaanasArora) [reply numpy/_core/src/multiarray/number.c]: Thanks for reviewing! It seems that in the old behavior only square was special-cased, as far as I can tell:

https://github.com/numpy/numpy/pull/27901/files#diff-1e638da588a4b86ad9e4c672ae59fa63c86a5232711d0a9f1a829ae89010d54aL464

It's a bit confusing, so would have been nice to have gotten rid of it, but well :)

> (If not, it may be nice to just use fastop != nops.square rather than introducing a new variable, but nitpicking.)

Actually yes, let me remove the variable, thanks!

Comment (seberg) [reply numpy/_core/src/multiarray/number.c]: Then maybe let's keep it at that for this PR... So we can consider backporting it (even if I am not sure I consider it a real regression).

**Comment by seberg** [comment]
Thanks @MaanasArora, let's put this in.

This should be simple enough to backport, it is a regression, but also a rather niche detail that I think is shaky to rely on.
